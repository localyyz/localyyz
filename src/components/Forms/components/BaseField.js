import React from "react";
import { View, StyleSheet, TouchableOpacity } from "react-native";

// third party
import PropTypes from "prop-types";
import MaterialIcon from "react-native-vector-icons/MaterialIcons";

// custom
import { Styles, Sizes, Colours } from "localyyz/constants";

// local
import Header from "./Header";
import Error from "./Error";

export default class BaseField extends React.Component {
  static propTypes = {
    isHorizontal: PropTypes.bool,
    label: PropTypes.string,
    hasError: PropTypes.bool,
    error: PropTypes.string
  };

  static defaultProps = {
    hasError: false,
    isHorizontal: false
  };

  get container() {
    return this.props.onPress ? TouchableOpacity : View;
  }

  get renderContent() {
    return (
      this.props.children
      || (this.props.onPress ? (
        <View style={styles.arrow}>
          <MaterialIcon
            name="keyboard-arrow-right"
            size={Sizes.Text}
            color={Colours.SubduedText}/>
        </View>
      ) : null)
    );
  }

  render() {
    return (
      <this.container
        onPress={this.props.onPress}
        style={[
          styles.container,
          this.props.style,
          this.props.hasMargin && styles.containerMargin
        ]}>
        <View
          style={[
            styles.wrapper,
            (this.props.isHorizontal || !this.props.children)
              && styles.horizontal
          ]}>
          <View style={styles.header}>
            {!this.props.hasError ? (
              <Header {...this.props}>{this.props.label}</Header>
            ) : (
              <Error>{this.props.error}</Error>
            )}
          </View>
          <View style={styles.content}>{this.renderContent}</View>
        </View>
      </this.container>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flexBasis: Math.min(Sizes.Width / 2, Sizes.InnerFrame * 10)
  },

  containerMargin: {
    marginVertical: Sizes.Spacer
  },

  wrapper: {
    paddingVertical: Sizes.InnerFrame,
    paddingHorizontal: Sizes.OuterFrame,
    backgroundColor: Colours.Foreground
  },

  horizontal: {
    ...Styles.Horizontal,
    ...Styles.EqualColumns
  },

  header: {
    marginRight: Sizes.InnerFrame
  },

  content: {
    flex: 1
  },

  arrow: {
    flex: 1,
    alignItems: "flex-end",
    justifyContent: "center"
  }
});
