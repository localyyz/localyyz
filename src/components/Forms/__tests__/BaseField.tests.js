import React from "react";
import Renderer from "react-test-renderer";

// local
import BaseField from "../components/BaseField";

// constants
const NAME = "Base Field test - form";
const PROPS = {
  label: "sample label",
  error: "sample error",
  onPress: jest.fn(),
  style: { backgroundColor: "red" }
};

const Component = BaseField;

describe(NAME, () => {
  it(`${NAME}: should render properly hasMargin:false`, () => {
    let props = { ...PROPS, onPress: jest.fn(), hasMargin: false };
    let comp = Renderer.create(<Component {...props} />);
    expect(comp.getInstance()).not.toBeUndefined();
    expect(comp.toJSON()).toMatchSnapshot();
  });

  it(`${NAME}: should render properly hasMargin:true`, () => {
    let props = { ...PROPS, onPress: jest.fn(), hasMargin: true };
    let comp = Renderer.create(<Component {...props} />);
    expect(comp.getInstance()).not.toBeUndefined();
    expect(comp.toJSON()).toMatchSnapshot();
  });

  it(`${NAME}: should render properly hasError:true`, () => {
    let props = {
      ...PROPS,
      onPress: jest.fn(),
      hasMargin: false,
      hasError: true
    };
    let comp = Renderer.create(<Component {...props} />);
    expect(comp.getInstance()).not.toBeUndefined();
    expect(comp.toJSON()).toMatchSnapshot();
  });

  it(`${NAME}: should render properly no onPress() given`, () => {
    let props = { ...PROPS, hasMargin: true, hasError: true };
    let comp = Renderer.create(<Component {...props} />);
    expect(comp.getInstance()).not.toBeUndefined();
    expect(comp.toJSON()).toMatchSnapshot();
  });

  it(`${NAME}: should render properly isHorizontal:false`, () => {
    let props = {
      ...PROPS,
      hasMargin: true,
      hasError: true,
      isHorizontal: false
    };
    let comp = Renderer.create(<Component {...props} />);
    expect(comp.getInstance()).not.toBeUndefined();
    expect(comp.toJSON()).toMatchSnapshot();
  });
});
