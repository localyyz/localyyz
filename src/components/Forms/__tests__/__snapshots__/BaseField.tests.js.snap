// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Base Field test - form Base Field test - form: should render properly hasError:true 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "red",
      "flexBasis": 150,
      "opacity": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#FFFFFF",
          "paddingHorizontal": 15,
          "paddingVertical": 10,
        },
        undefined,
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
        },
      ]
    }
  >
    <View
      style={
        Object {
          "marginRight": 15,
        }
      }
    >
      <View
        style={Object {}}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          numberOfLines={1}
          style={
            Object {
              "backgroundColor": "rgba(255, 255, 255, 0)",
              "color": "#FF0000",
              "fontSize": 14,
              "fontWeight": "300",
            }
          }
        >
          sample error
        </Text>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
          },
          undefined,
        ]
      }
    >
      <View
        style={
          Object {
            "alignItems": "flex-end",
            "flex": 1,
            "justifyContent": "center",
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={false}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#999999",
                "fontSize": 16,
              },
              undefined,
              Object {
                "fontFamily": "Material Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
            ]
          }
        >
          
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`Base Field test - form Base Field test - form: should render properly hasMargin:false 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "red",
      "flexBasis": 150,
      "opacity": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#FFFFFF",
          "paddingHorizontal": 15,
          "paddingVertical": 10,
        },
        undefined,
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
        },
      ]
    }
  >
    <View
      style={
        Object {
          "marginRight": 15,
        }
      }
    >
      <View
        style={Object {}}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          numberOfLines={1}
          style={
            Array [
              Object {
                "backgroundColor": "rgba(255, 255, 255, 0)",
                "color": "rgb(30, 30, 30)",
                "fontSize": 14,
                "fontWeight": "300",
              },
              undefined,
            ]
          }
        >
          sample label
        </Text>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
          },
          undefined,
        ]
      }
    >
      <View
        style={
          Object {
            "alignItems": "flex-end",
            "flex": 1,
            "justifyContent": "center",
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={false}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#999999",
                "fontSize": 16,
              },
              undefined,
              Object {
                "fontFamily": "Material Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
            ]
          }
        >
          
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`Base Field test - form Base Field test - form: should render properly hasMargin:true 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "red",
      "flexBasis": 150,
      "marginVertical": 0.5,
      "opacity": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#FFFFFF",
          "paddingHorizontal": 15,
          "paddingVertical": 10,
        },
        undefined,
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
        },
      ]
    }
  >
    <View
      style={
        Object {
          "marginRight": 15,
        }
      }
    >
      <View
        style={Object {}}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          numberOfLines={1}
          style={
            Array [
              Object {
                "backgroundColor": "rgba(255, 255, 255, 0)",
                "color": "rgb(30, 30, 30)",
                "fontSize": 14,
                "fontWeight": "300",
              },
              undefined,
            ]
          }
        >
          sample label
        </Text>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
          },
          undefined,
        ]
      }
    >
      <View
        style={
          Object {
            "alignItems": "flex-end",
            "flex": 1,
            "justifyContent": "center",
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={false}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#999999",
                "fontSize": 16,
              },
              undefined,
              Object {
                "fontFamily": "Material Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
            ]
          }
        >
          
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`Base Field test - form Base Field test - form: should render properly isHorizontal:false 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "red",
      "flexBasis": 150,
      "marginVertical": 0.5,
      "opacity": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#FFFFFF",
          "paddingHorizontal": 15,
          "paddingVertical": 10,
        },
        undefined,
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
        },
      ]
    }
  >
    <View
      style={
        Object {
          "marginRight": 15,
        }
      }
    >
      <View
        style={Object {}}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          numberOfLines={1}
          style={
            Object {
              "backgroundColor": "rgba(255, 255, 255, 0)",
              "color": "#FF0000",
              "fontSize": 14,
              "fontWeight": "300",
            }
          }
        >
          sample error
        </Text>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
          },
          undefined,
        ]
      }
    >
      <View
        style={
          Object {
            "alignItems": "flex-end",
            "flex": 1,
            "justifyContent": "center",
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={false}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#999999",
                "fontSize": 16,
              },
              undefined,
              Object {
                "fontFamily": "Material Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
            ]
          }
        >
          
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`Base Field test - form Base Field test - form: should render properly no onPress() given 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "red",
      "flexBasis": 150,
      "marginVertical": 0.5,
      "opacity": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#FFFFFF",
          "paddingHorizontal": 15,
          "paddingVertical": 10,
        },
        undefined,
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
        },
      ]
    }
  >
    <View
      style={
        Object {
          "marginRight": 15,
        }
      }
    >
      <View
        style={Object {}}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          numberOfLines={1}
          style={
            Object {
              "backgroundColor": "rgba(255, 255, 255, 0)",
              "color": "#FF0000",
              "fontSize": 14,
              "fontWeight": "300",
            }
          }
        >
          sample error
        </Text>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
          },
          undefined,
        ]
      }
    >
      <View
        style={
          Object {
            "alignItems": "flex-end",
            "flex": 1,
            "justifyContent": "center",
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={false}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#999999",
                "fontSize": 16,
              },
              undefined,
              Object {
                "fontFamily": "Material Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
            ]
          }
        >
          
        </Text>
      </View>
    </View>
  </View>
</View>
`;
